### DISCUSSION TITLE
Update the latest value in ITEM_INPUT

### DISCUSSION THREAD

**Discussion by Aditya-Shandilya** [forntoh/LcdMenu#144] [Q&A] [answered]
Hello, I am new to this library and I am working on a project that involves calculating the area of either a rectangle or a circle. I have three variables: Length, Breadth, and Diameter. To facilitate the calculation, I've set up a system where I can input values for these dimensions. However, I've encountered a problem where entering a value for Length works fine, but when I input a value for Breadth, it seems to overwrite the Length value. The same issue occurs with Diameter. I debugged it a bit and found out that it is successfully storing the inputs in the variable but not updating it on the Lcd Screen rather the value that is being inputted globally is updated. Heres the snippet of my code:

```
#define CHARSET_SIZE 10
char charset[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
uint8_t charsetPosition;

int length = 0;
int breadth = 0;
int diameter = 0;

bool lengthSet = false;
bool breadthSet = false;
bool diameterSet = false;

void lengthInputCallback(char *value);
void breadthInputCallback(char *value);
void diameterInputCallback(char *value);

extern MenuItem *AreaMenu[];
extern MenuItem *RectangleSubMenu[];
extern MenuItem *CircleSubMenu[];

MAIN_MENU(
    ITEM_SUBMENU("Set Area", AreaMenu));

SUB_MENU(AreaMenu, mainMenu,
         ITEM_SUBMENU("Rectangle", RectangleSubMenu),
         ITEM_SUBMENU("Circle", CircleSubMenu));

SUB_MENU(RectangleSubMenu, AreaMenu,
         ITEM_INPUT("Length", lengthInputCallback),
         ITEM_INPUT("Breadth", breadthInputCallback));

SUB_MENU(CircleSubMenu, AreaMenu,
         ITEM_INPUT("Diameter", diameterInputCallback));
```
Above is the definition for my menu. 
Button Definition that was referred from the example of the library:

```
if (debounce(BUTTON_UP))
  {
    if (menu.isInEditMode()) // Update the position only in edit mode
      charsetPosition = (charsetPosition + 1) % CHARSET_SIZE;
    menu.drawChar(charset[charsetPosition]); // Works only in edit mode
    menu.up();
  }
  else if (debounce(BUTTON_DOWN))
  {
    if (menu.isInEditMode())
      charsetPosition =
          constrain(charsetPosition - 1, 0, CHARSET_SIZE);
    menu.drawChar(charset[charsetPosition]);
    menu.down();
  }
  else if (debounce(BUTTON_ENTER))
  {
    if (menu.isInEditMode())
    {
      menu.type(charset[charsetPosition]);
    }
    charsetPosition = 0;
    menu.enter();
  }
  else if (debounce(BUTTON_BACK))
  {
    menu.back();
  }
}
```

Converting the string to an integer and storing it in a variable:

```
void lengthInputCallback(char *value)
{
  length = atoi(value); // Store the length value in the variable.
  lengthSet = true;
}
void breadthInputCallback(char *value)
{
  breadth = atoi(value);
  breadthSet = true;
}
void diameterInputCallback(char *value)
{
  diameter = atoi(value);
  diameterSet = true;
}
```

I have referred to the ItemInput example for character sets and managed to input values accordingly. However, the issue with values overwriting each other remains unresolved. Note that I have checked that all my connections are correct and the buttons are pressed correctly. The code also runs without any errors. Any guidance or suggestions to correct this behavior would be greatly appreciated.

**Comment by forntoh** [comment]
I added a few helpers in [v3.5.0](https://github.com/forntoh/LcdMenu/releases/tag/3.5.0), they should help you.

Your code could become something like this

```cpp
processMenuCommand(
    menu,
    command,
    debounce(BUTTON_UP),
    debounce(BUTTON_DOWN),
    debounce(BUTTON_ENTER),
    debounce(BUTTON_BACK)
);
```

I don't know what your debounce function does/returns but in theory, it should work. I'd prefer to use a function that reads the button and returns the current state/action to perform.

e.g. https://forum.arduino.cc/t/debounced-button-reading-help/599533/6

**Comment by Engi-xm** [comment]
I ran into the same issue. I don't think it has anything to do with debounce or using processMenuCommand.
The problem is that when you have two ITEM_INPUT objects in the same menu screen, the value of the first one is visually overwritten by the second one. The actual value seems to be fine, it's only a visual bug on the LCD.
I slightly modified the provided example SimpleInput which recreates the problem:

```
/*
 Simple Input with visual bug
*/

#include <ItemInput.h>
#include <LcdMenu.h>
#include <utils/commandProccesors.h>

#define LCD_ROWS 4
#define LCD_COLS 20

// Configure keyboard keys (ASCII)
#define UP 56        // NUMPAD 8
#define DOWN 50      // NUMPAD 2
#define LEFT 52      // NUMPAD 4
#define RIGHT 54     // NUMPAD 6
#define ENTER 53     // NUMPAD 5
#define BACK 55      // NUMPAD 7
#define BACKSPACE 8  // BACKSPACE
#define CLEAR 46     // NUMPAD .

// Declare the call back function
void inputCallback(char* value);
void inputCallback2(char* value);

MAIN_MENU(
    ITEM_INPUT("Con", inputCallback), 
    ITEM_INPUT("wifi", inputCallback2),
    ITEM_BASIC("Blink SOS"), 
    ITEM_BASIC("Blink random")
);

LcdMenu menu(LCD_ROWS, LCD_COLS);

void setup() {
    Serial.begin(9600);
    menu.setupLcdWithMenu(0x3F, mainMenu);
}

void loop() {
    if (!Serial.available()) return;
    char command = Serial.read();

    processMenuCommand(menu, command, UP, DOWN, LEFT, RIGHT, ENTER, BACK, CLEAR,
                       BACKSPACE);
}
/**
 * Define callback
 */
void inputCallback(char* value) {
    Serial.print(F("# "));
    Serial.println(value);
}
void inputCallback2(char* value) {
    Serial.print(F("# "));
    Serial.println(value);
}
```

I managed to find a workaround, but to be honest have no idea why it works. I use two different arrays as initial values for the two ITEM_INPUT objects:

```
char val[5] = {' ', ' ', ' ', ' ', '\0'};
char val2[5]= {'\0', '\0', '\0', '\0', '\0'};

MAIN_MENU(
    ITEM_INPUT("Con", val, inputCallback), 
    ITEM_INPUT("wifi", val2, inputCallback2),
    ITEM_BASIC("Blink SOS"), 
    ITEM_BASIC("Blink random")
);
```

**Reply by forntoh** [reply]
It has something to do with memory allocation.
That's the reason why it doesn't work with one array, because you'll be copying the address of the same element, and a new memory location is not allocated.

**Comment by Engi-xm** [comment]
I don't think so. I probably wasn't clear enough with my workaround. The two arrays are different in content, not only names (instances).
For example this DOES NOT work:
```
char val[5];
char val2[5];

MAIN_MENU(
    ITEM_INPUT("Con", val, inputCallback), 
    ITEM_INPUT("wifi", val2, inputCallback2),
    ITEM_BASIC("Blink SOS"), 
    ITEM_BASIC("Blink random")
);
```
This also DOES NOT work:
```
char val[5] = {''\0', '\0', '\0', '\0', '\0'};
char val2[5]= {'\0', '\0', '\0', '\0', '\0'};

MAIN_MENU(
    ITEM_INPUT("Con", val, inputCallback), 
    ITEM_INPUT("wifi", val2, inputCallback2),
    ITEM_BASIC("Blink SOS"), 
    ITEM_BASIC("Blink random")
);
```
Notice that the contents of the two arrays are different in the example that does work (one is filled with '\0', the other with '<space>').

**Comment by forntoh** [comment, marked as answer]
## Fixed in v3.5.2 🚀
Issue #158
